analyzer/src/steps/typing/exploration.rs

```diff
@@ -1,34 +1,77 @@
-use context::source::SourceSegment;
-
-use crate::diagnostic::{Diagnostic, DiagnosticID, Observation};
-use crate::relations::SourceId;
+use crate::engine::Engine;
+use crate::reef::{ReefId, Reefs};
+use crate::relations::Relations;
 use crate::steps::typing::function::Return;
 use crate::types::ctx::TypeContext;
 use crate::types::engine::TypedEngine;
-use crate::types::hir::TypeId;
-use crate::types::ty::Type;
 use crate::types::Typing;
 
 /// The support for type analysis.
 pub(super) struct Exploration {
-    pub(super) engine: TypedEngine,
+    pub(super) type_engine: TypedEngine,
     pub(super) typing: Typing,
     pub(super) ctx: TypeContext,
     pub(super) returns: Vec<Return>,
 }
 
+#[derive(Clone)]
+pub struct ReefTypes<'a> {
+    pub context: &'a TypeContext,
+    pub engine: &'a TypedEngine,
+    pub typing: &'a Typing,
+}
+
+/// A proxy that let the user access to any reef's engine, relations and types data.
+/// If the given reef identifier is the `current_reef`, the types data stored in `current` is returned instead
+/// of the data stored by the `reefs`.
+pub struct UniversalReefAccessor<'a, 'e> {
```

Comment:
I have doubts about this structure that only allow immutable access. If it is really immutable, mark it `Clone` and `Copy`.

My concern here is that it impose immutable access everywhere, where the current `Exploration` is mutable.